botocore/client.py

```diff
@@ -114,21 +115,50 @@ def _get_signature_version_and_region(self, service_model, region_name,
                         service_model.endpoint_prefix, override)
                     signature_version = override
 
+        # Determine the region name as well
+        region_name = self._determine_region_name(
+            endpoint_config, region_name, endpoint_url)
+
         return signature_version, region_name
 
+    def _determine_region_name(self, endpoint_config, region_name=None,
+                               endpoint_url=None):
+        # This is a helper function to determine region name to use.
+        # It will take into account whether the user passes in a region
+        # name, whether their is a rule in the endpoint JSON, or
+        # an endpoint url was provided.
+
+        # We only support the credentialScope.region in the properties
+        # bag right now, so if it's available, it will override the
+        # provided region name.
+        region_name_override = endpoint_config['properties'].get(
+            'credentialScope', {}).get('region')
+
+        if endpoint_url is not None:
+            # If an endpoint_url is provided, do not use region name
+            # override if a region
+            # was provided by the user.
```

Comment:
This comment looks split onto a new line kind of strange. It could just be two lines instead of three.